### PR TITLE
[DESK] Find best display settings

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Based on KRosUser's `desk.patch`.
JIRA issue: [CORE-18187](https://jira.reactos.org/browse/CORE-18187), [CORE-16878](https://jira.reactos.org/browse/CORE-16878)

## Proposed changes

- Add `FindBestElement` helper function.
- Use `FindBestElement` to find the best display settings.

## TODO

- [x] Do tests.

### CODE REVIEW

File: dll/cpl/desk/settings.c [hunk 1]
```diff
@@ -842,6 +843,42 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+        Current = Current->Flink;
+    }
+
+    if ((Current->dmPelsWidth == dmPelsWidth) &&
+        (Current->dmPelsHeight == dmPelsHeight) &&
+        (Current->dmPelsHeight == dmPelsHeight) &&
+        (Current->dmBitsPerPel == dmBitsPerPel))
+    {
+        return Current; /* Found */
+    }
```
Comment (binarymaster) [inline dll/cpl/desk/settings.c:877]:
```suggestion
    while (1)
    {
        if ((Current->dmPelsWidth == dmPelsWidth) &&
            (Current->dmPelsHeight == dmPelsHeight) &&
            (Current->dmPelsHeight == dmPelsHeight) &&
            (Current->dmBitsPerPel == dmBitsPerPel))
        {
            return Current; /* Found */
        }

        if (!Current->Flink)
            break;

        Current = Current->Flink;
    }
```

File: dll/cpl/desk/settings.c [hunk 2]
```diff
@@ -842,6 +843,37 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+PSETTINGS_ENTRY
+FindBestElement(DWORD dmPelsWidth, DWORD dmPelsHeight, DWORD dmBitsPerPel, PSETTINGS_ENTRY List)
+{
+    PSETTINGS_ENTRY Current = List;
+
+    /* Go back to 1st element */
+    while (Current->Blink)
+    {
+        Current = Current->Blink;
+    }
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:855]:
```suggestion
        Current = Current->Blink;
```

File: dll/cpl/desk/settings.c [hunk 3]
```diff
@@ -842,6 +843,37 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+
+    /* Go through the list */
+    while (1)
+    {
+        if ((Current->dmPelsWidth == dmPelsWidth) &&
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:860]:
```suggestion
        if (Current->dmPelsWidth == dmPelsWidth &&
[...]
```

File: dll/cpl/desk/settings.c [hunk 4]
```diff
@@ -842,6 +843,37 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+            (Current->dmPelsHeight == dmPelsHeight) &&
+            (Current->dmPelsHeight == dmPelsHeight) &&
+            (Current->dmBitsPerPel == dmBitsPerPel))
+        {
+            return Current; /* Found */
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:865]:
```suggestion
            return Current;
```

File: dll/cpl/desk/settings.c [hunk 5]
```diff
@@ -842,6 +843,37 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+        }
+
+        if (!Current->Flink)
+            break;
+
+        Current = Current->Flink;
+    }
+
+    return NULL; /* Not found */
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:874]:
```suggestion
    return NULL;
```

File: dll/cpl/desk/settings.c [hunk 6]
```diff
@@ -866,10 +898,14 @@ ApplyDisplaySettings(HWND hwndDlg, PSETTINGS_DATA pData)
...
-        pData->CurrentDisplayDevice->CurrentSettings->dmBitsPerPel = pData->CurrentDisplayDevice->InitialSettings.dmBitsPerPel;
-        pData->CurrentDisplayDevice->CurrentSettings->dmDisplayFrequency = pData->CurrentDisplayDevice->InitialSettings.dmDisplayFrequency;
+        PSETTINGS_ENTRY pInitialSettings = &pData->CurrentDisplayDevice->InitialSettings;
+        pData->CurrentDisplayDevice->CurrentSettings =
+            FindBestElement(pInitialSettings->dmPelsWidth,
+                            pInitialSettings->dmPelsHeight,
+                            pInitialSettings->dmBitsPerPel,
+                            pData->CurrentDisplayDevice->CurrentSettings);
+        /* Initial entry should be found */
+        ASSERT(pData->CurrentDisplayDevice->CurrentSettings);
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:908]: Unless some other usages are planned, move `ASSERT()` to the function.

File: dll/cpl/desk/settings.c [hunk 7]
```diff
@@ -866,10 +898,14 @@ ApplyDisplaySettings(HWND hwndDlg, PSETTINGS_DATA pData)
...
     {
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsWidth = pData->CurrentDisplayDevice->InitialSettings.dmPelsWidth;
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsHeight = pData->CurrentDisplayDevice->InitialSettings.dmPelsHeight;
-        pData->CurrentDisplayDevice->CurrentSettings->dmBitsPerPel = pData->CurrentDisplayDevice->InitialSettings.dmBitsPerPel;
-        pData->CurrentDisplayDevice->CurrentSettings->dmDisplayFrequency = pData->CurrentDisplayDevice->InitialSettings.dmDisplayFrequency;
+        PSETTINGS_ENTRY pInitialSettings = &pData->CurrentDisplayDevice->InitialSettings;
+        pData->CurrentDisplayDevice->CurrentSettings =
+            FindBestElement(pInitialSettings->dmPelsWidth,
+                            pInitialSettings->dmPelsHeight,
+                            pInitialSettings->dmBitsPerPel,
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/cpl/desk/settings.c:902]: frequency is not so important.

File: dll/cpl/desk/settings.c [hunk 8]
```diff
@@ -866,10 +898,14 @@ ApplyDisplaySettings(HWND hwndDlg, PSETTINGS_DATA pData)
     }
     else
     {
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsWidth = pData->CurrentDisplayDevice->InitialSettings.dmPelsWidth;
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsHeight = pData->CurrentDisplayDevice->InitialSettings.dmPelsHeight;
-        pData->CurrentDisplayDevice->CurrentSettings->dmBitsPerPel = pData->CurrentDisplayDevice->InitialSettings.dmBitsPerPel;
-        pData->CurrentDisplayDevice->CurrentSettings->dmDisplayFrequency = pData->CurrentDisplayDevice->InitialSettings.dmDisplayFrequency;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/cpl/desk/settings.c:872]: frequency is not so important.

File: dll/cpl/desk/settings.c (hunk 6 above)
Comment (katahiromz) [reply dll/cpl/desk/settings.c]: Moved.

File: dll/cpl/desk/settings.c [hunk 9]
```diff
@@ -842,6 +844,43 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+                  0x100 * labs(Current->dmPelsHeight - pInitialSettings->dmPelsHeight) +
+                  labs(Current->dmDisplayFrequency - pInitialSettings->dmDisplayFrequency);
+        if (Penalty == 0)
+            return Current;
+
+        if (!pBestEntry || SmallestPenalty > Penalty)
+        {
+            SmallestPenalty = Penalty;
+            pBestEntry = Current;
+        }
```
Comment (binarymaster) [inline dll/cpl/desk/settings.c:874]: Do we need such complex maths for that?

File: dll/cpl/desk/settings.c [hunk 10]
```diff
@@ -9,6 +9,7 @@
  */
 
 #include "desk.h"
+#include <debug.h>
```
[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/cpl/desk/settings.c:13]: If this is used just for ASSERT, I would recommend instead using <assert.h> and the similarly-named macro in lowercase.

File: dll/cpl/desk/settings.c [hunk 11]
```diff
@@ -842,6 +844,43 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+FindBestElement(
+    _In_ PSETTINGS_ENTRY pInitialSettings,
+    _In_ PSETTINGS_ENTRY Current)
+{
+    LONG Penalty, SmallestPenalty = MAXLONG;
+    PSETTINGS_ENTRY pBestEntry = NULL;
+
+    /* Go back to 1st element */
+    while (Current->Blink)
+        Current = Current->Blink;
```
Comment (HBelusca) [inline dll/cpl/desk/settings.c:858]: I wonder whether giving to the FindBestElement function also the head of this linked list wouldn't be better.

File: dll/cpl/desk/settings.c (hunk 9 above)
Comment (katahiromz) [reply dll/cpl/desk/settings.c:874]: If the perfect choice didn't exist, then it will be the nearest choice.

File: dll/cpl/desk/settings.c (hunk 11 above)
Comment (katahiromz) [reply dll/cpl/desk/settings.c]: The head node is used. Thanks.

File: dll/cpl/desk/settings.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/cpl/desk/settings.c]: Used `<assert.h>`.

File: dll/cpl/desk/settings.c [hunk 12]
```diff
@@ -866,10 +899,9 @@ ApplyDisplaySettings(HWND hwndDlg, PSETTINGS_DATA pData)
     }
     else
     {
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsWidth = pData->CurrentDisplayDevice->InitialSettings.dmPelsWidth;
-        pData->CurrentDisplayDevice->CurrentSettings->dmPelsHeight = pData->CurrentDisplayDevice->InitialSettings.dmPelsHeight;
-        pData->CurrentDisplayDevice->CurrentSettings->dmBitsPerPel = pData->CurrentDisplayDevice->InitialSettings.dmBitsPerPel;
-        pData->CurrentDisplayDevice->CurrentSettings->dmDisplayFrequency = pData->CurrentDisplayDevice->InitialSettings.dmDisplayFrequency;
+        PSETTINGS_ENTRY pInitialSettings = &pData->CurrentDisplayDevice->InitialSettings;
+        pData->CurrentDisplayDevice->CurrentSettings =
+            FindBestElement(pInitialSettings, pData->CurrentDisplayDevice->Settings);
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:904]:
```suggestion
        pData->CurrentDisplayDevice->CurrentSettings =
            FindBestElement(&pData->CurrentDisplayDevice->InitialSettings,
                            pData->CurrentDisplayDevice->Settings);
```

--

Going a step further, you could pass `pData->CurrentDisplayDevice` only.

File: dll/cpl/desk/settings.c [hunk 13]
```diff
@@ -842,6 +844,37 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+            return Current;
+
+        if (Distance < NearestDistance)
+        {
+            pBestEntry = Current;
+            NearestDistance = Distance;
+        }
+    }
+
+    assert(pBestEntry);
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:873]: Not needed anymore, as I assume `List` cannot be empty, can it?

File: dll/cpl/desk/settings.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply dll/cpl/desk/settings.c]: `dmDisplayFrequency` now taken into account too.

**Comment by HBelusca** [comment]
Cc @hpoussin for some review on this PR

File: dll/cpl/desk/settings.c (hunk 12 above)
Comment (katahiromz) [reply dll/cpl/desk/settings.c]: Thanks.

File: dll/cpl/desk/settings.c (hunk 13 above)
Comment (katahiromz) [reply dll/cpl/desk/settings.c]: Deleted.

File: dll/cpl/desk/settings.c [hunk 14]
```diff
@@ -842,6 +842,35 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
...
+PSETTINGS_ENTRY
+FindBestElement(
+    _In_ PDISPLAY_DEVICE_ENTRY pDevice)
+{
+    PSETTINGS_ENTRY Current, Request = &pDevice->InitialSettings, List = pDevice->Settings;
+    PSETTINGS_ENTRY pBestEntry = NULL;
+    LONG Distance, NearestDistance = MAXLONG;
+
+    /* Find the best entry in the list */
+    for (Current = List; Current; Current = Current->Flink)
```
Comment (SergeGautherie) [inline dll/cpl/desk/settings.c:855]:
```suggestion
    PSETTINGS_ENTRY Request = &pDevice->InitialSettings, pBestEntry = NULL, Current;
    LONG Distance, NearestDistance = MAXLONG;

    /* Find the best entry in the list */
    for (Current = pDevice->Settings; Current; Current = Current->Flink)
```

--

And you may `s/pBestEntry/BestEntry/g`.

**Comment by katahiromz** [comment]
KRosUser tested successfully. Should be okay.

**Pull request merged**